### ISSUE TITLE
Consider using Waku protocol

### ISSUE THREAD

**Issue by ameba23** [entropyxyz/entropy-core#584]
This follows a discussion at the San Diego off-site.

## What is it?

[Waku](https://waku.org) is set of protocols designed for secure messaging, used by the [Status](https://status.app) messaging app as well as [the Graph](https://thegraph.com) Ethereum indexer. 

The [Waku Relay](https://rfc.vac.dev/spec/11) protocol is based on the [Libp2p Gossipsub protocol](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/README.md).

Anyone running the protocol can publish a message to a specified topic name, and the message will be 'gossiped' to all nodes running the protocol. There are various privacy features such as publisher/subscriber unlinkability described in the protocol spec. 

'Light clients' (eg: browser or mobile users) are able to publish messages as well receive messages from subscribed topics without actively participating in message distribution, through the [Light Push](https://rfc.vac.dev/spec/19) and [Waku Filter](https://rfc.vac.dev/spec/12) protocols.

Generally peer discovery is done through one or more hard-coded bootstrap nodes.

Libp2p is transport agnostic, so the protocol can theoretically run on different transports but generally it uses TCP/websockets. 

Payload encryption can be added through the [waku2-noise](https://rfc.vac.dev/spec/35) protocol.

AFAIK there is currently no pure rust implementation, only [rust bindings to the Go implementation](https://crates.io/crates/waku-bindings). But since the relay protocol is basically just libp2p's gossipsub (for which there [is a pure rust implementation](https://docs.rs/libp2p/0.53.2/libp2p/gossipsub/index.html)) with a [custom protobuf message type](https://rfc.vac.dev/spec/14/#wire-format), it is easy to use from rust anyway.

## Why would we want to use it?

I think for TSS node-to-node communication it does not really improve at all on what we already have. Its anonymity features are not needed since node IPs are anyway public.

But for node-to-user communication it could be useful, as users would not have to connect directly with TSS servers, they could relay message through other waku nodes, connecting only to nodes which use TLS, and therefore being able to connect even in a secure context in the browser, which is not possible with our current setup.

I think for signing in public or permissioned access mode, this would work really well, as there are only two messages needed to get a signature. The user would publish a signature request, which would be encrypted such that it is only accessible to the TSS nodes in the signing committee for that particular message. It would be relayed over the Waku/Entropy network. The signing protocol would be performed just as it is now (not using Waku). And then the resulting signature or error message would be encrypted for the user and relayed back to them again using the Waku Relay protocol.

This is essentially very similar to having a proxy server to relay messages between the user and TSS servers, which we already talked about doing. The difference is we get redundancy because there is not just one proxy server but potentially many, and we can rely on existing infrastructure provided by the Waku network as well as our own.

However for private access mode i am less sure it is a good option, because the DKG and Signing protocols involve exchanging many messages, and the extra latency of relaying messages will have a bigger impact.

## TLDR

To summarize, i think using the Waku Relay protocol could be useful for building an Entropy client which needs to work from a secure context in the browser. For private access mode it is less suitable, but there are anyway other issues related to doing private access mode in the browser, such as secure storage of the key-share.

**Comment by ameba23** [comment]
Closing this as i am no longer considering it. Feel free to re-open.
